File: packages/pdf-gen-py/pdf_gen/templates/invoice.py
from typing import Any, Dict

from ..config import PDFConfig, Theme
from ..surface import Surface
from ..utils.table import draw_table
from ..utils.validate import require_fields, format_date, format_currency


def render_invoice(surface: Surface, config: PDFConfig, data: Dict[str, Any], theme: Theme) -> None:
    require_fields(data, ['invoice_number', 'company_name', 'client_name', 'items', 'subtotal', 'total'], 'invoice')

    left = config.margins['left']
    right = surface.page_width - config.margins['right']
    content_width = right - left
    currency = data.get('currency', 'USD')

    # --- Header banner -------------------------------------------------
    banner_height = 90
    surface.rect(0, 0, surface.page_width, banner_height, fill=theme.primary)

    surface.text(left, 30, 'INVOICE', font='Helvetica-Bold', size=26, color='#ffffff')
    surface.text(left, 62, data['company_name'], font='Helvetica', size=10, color='#ffffff', width=content_width * 0.6)

    surface.text(
        left, 30, f"#{data['invoice_number']}", font='Helvetica-Bold', size=12,
        color='#ffffff', align='right', width=content_width
    )
    surface.text(
        left, 50, format_date(data.get('date')), font='Helvetica', size=10,
        color='#ffffff', align='right', width=content_width
    )

    status = data.get('status')
    if status:
        status_colors = {'PAID': '#1e8e3e', 'DUE': theme.accent, 'OVERDUE': '#c62828'}
        surface.text(
            left, 68, status, font='Helvetica-Bold', size=10,
            color=status_colors.get(status, theme.accent), align='right', width=content_width
        )

    # --- From / Bill To --------------------------------------------------
    y = banner_height + 30
    surface.text(left, y, 'FROM', font='Helvetica-Bold', size=9, color=theme.muted)
    surface.text(left + content_width / 2, y, 'BILL TO', font='Helvetica-Bold', size=9, color=theme.muted)

    y += 14
    surface.text(left, y, data['company_name'], font='Helvetica-Bold', size=11, color=theme.text, width=content_width / 2 - 10)
    surface.text(left + content_width / 2, y, data['client_name'], font='Helvetica-Bold', size=11, color=theme.text, width=content_width / 2)

    y += 16
    company_address = data.get('company_address')
    client_address = data.get('client_address')
    if company_address:
        surface.text(left, y, company_address, font='Helvetica', size=10, color=theme.muted, width=content_width / 2 - 10)
    if client_address:
        surface.text(left + content_width / 2, y, client_address, font='Helvetica', size=10, color=theme.muted, width=content_width / 2)

    y += 20
    if data.get('due_date'):
        surface.text(left, y, f"Due date: {format_date(data['due_date'])}", font='Helvetica-Bold', size=9, color=theme.muted)
        y += 10

    # --- Line items table --------------------------------------------------
    table_top = y + 15
    rows = []
    for item in data['items']:
        item_total = item.get('total', item['quantity'] * item['unit_price'])
        rows.append({
            'description': item['description'],
            'quantity': str(item['quantity']),
            'unit_price': format_currency(item['unit_price'], currency),
            'total': format_currency(item_total, currency),
        })

    final_y = draw_table(
        surface, config, theme,
        columns=[
            {'header': 'Description', 'key': 'description'},
            {'header': 'Qty', 'key': 'quantity', 'width': content_width * 0.12, 'align': 'right'},
            {'header': 'Unit Price', 'key': 'unit_price', 'width': content_width * 0.2, 'align': 'right'},
            {'header': 'Total', 'key': 'total', 'width': content_width * 0.2, 'align': 'right'},
        ],
        rows=rows,
        start_y=table_top,
    )

    # --- Totals summary --------------------------------------------------
    summary_width = 220
    summary_x = right - summary_width
    sy = final_y + 15

    surface.text(summary_x, sy, 'Subtotal', font='Helvetica', size=10, color=theme.text, width=summary_width - 90)
    surface.text(summary_x + summary_width - 90, sy, format_currency(data['subtotal'], currency), font='Helvetica', size=10, color=theme.text, align='right', width=90)
    sy += 18

    tax = data.get('tax')
    tax_rate = data.get('tax_rate')
    if tax or tax_rate:
        tax_label = f'Tax ({tax_rate}%)' if tax_rate else 'Tax'
        surface.text(summary_x, sy, tax_label, font='Helvetica', size=10, color=theme.text, width=summary_width - 90)
        surface.text(summary_x + summary_width - 90, sy, format_currency(tax or 0, currency), font='Helvetica', size=10, color=theme.text, align='right', width=90)
        sy += 18

    surface.line(summary_x, sy, summary_x + summary_width, sy, color=theme.border)
    sy += 8

    surface.rect(summary_x, sy, summary_width, 30, fill=theme.primary)
    surface.text(summary_x + 10, sy + 8, 'Total', font='Helvetica-Bold', size=12, color='#ffffff', width=summary_width - 100)
    surface.text(summary_x + summary_width - 100, sy + 8, format_currency(data['total'], currency), font='Helvetica-Bold', size=12, color='#ffffff', align='right', width=90)

    # --- Notes & payment terms --------------------------------------------------
    notes = data.get('notes')
    payment_terms = data.get('payment_terms')
    if notes or payment_terms:
        ny = sy + 55
        if ny + 60 > surface.page_height - config.margins['bottom']:
            surface.new_page()
            ny = config.margins['top']

        surface.text(left, ny, 'NOTES', font='Helvetica-Bold', size=9, color=theme.muted)
        ny += 14
        if notes:
            ny = surface.text(left, ny, notes, font='Helvetica', size=10, color=theme.text, width=content_width)
        if payment_terms:
            surface.text(left, ny + 4, f'Payment terms: {payment_terms}', font='Helvetica', size=10, color=theme.text, width=content_width)

File: packages/pdf-gen-js/src/templates/invoice.ts
import PDFDocument from 'pdfkit';
import { PDFConfig } from '../config';

export interface InvoiceData {
  invoiceNumber: string;
  date: Date;
  companyName: string;
  companyAddress?: string;
  clientName: string;
  clientAddress?: string;
  items: Array<{
    description: string;
    quantity: number;
    unitPrice: number;
    total?: number;
  }>;
  subtotal: number;
  tax?: number;
  total: number;
  notes?: string;
  paymentTerms?: string;
}

export function renderInvoice(pdf: PDFDocument, config: PDFConfig, data: InvoiceData): void {
  const leftMargin = config.margins.left;
  const topMargin = config.margins.top;
  const pageWidth = pdf.page.width - config.margins.left - config.margins.right;

  pdf.font('Helvetica-Bold', 24).text('INVOICE', leftMargin, topMargin);

  pdf.font('Helvetica', 10).text(`Invoice #: ${data.invoiceNumber}`, leftMargin, topMargin + 40);
  pdf.text(`Date: ${data.date.toLocaleDateString()}`);

  pdf.font('Helvetica-Bold', 12).text('From:', leftMargin, topMargin + 80);
  pdf.font('Helvetica', 11).text(data.companyName);
  if (data.companyAddress) pdf.text(data.companyAddress);

  pdf.font('Helvetica-Bold', 12).text('Bill To:', leftMargin + 300, topMargin + 80);
  pdf.font('Helvetica', 11).text(data.clientName, leftMargin + 300);
  if (data.clientAddress) pdf.text(data.clientAddress, leftMargin + 300);

  const tableTop = topMargin + 160;
  const columns = ['Description', 'Quantity', 'Unit Price', 'Total'];
  const columnWidths = [pageWidth * 0.5, pageWidth * 0.15, pageWidth * 0.15, pageWidth * 0.2];

  pdf.font('Helvetica-Bold', 11);
  let xPos = leftMargin;
  columns.forEach((col, i) => {
    pdf.text(col, xPos, tableTop);
    xPos += columnWidths[i];
  });

  pdf.moveTo(leftMargin, tableTop + 20).lineTo(leftMargin + pageWidth, tableTop + 20).stroke();

  pdf.font('Helvetica', 10);
  let yPos = tableTop + 30;
  data.items.forEach((item) => {
    const itemTotal = item.total || item.quantity * item.unitPrice;
    xPos = leftMargin;
    pdf.text(item.description, xPos, yPos, { width: columnWidths[0] });
    xPos += columnWidths[0];
    pdf.text(item.quantity.toString(), xPos, yPos, { width: columnWidths[1] });
    xPos += columnWidths[1];
    pdf.text(`$${item.unitPrice.toFixed(2)}`, xPos, yPos, { width: columnWidths[2] });
    xPos += columnWidths[2];
    pdf.text(`$${itemTotal.toFixed(2)}`, xPos, yPos, { width: columnWidths[3] });
    yPos += 20;
  });

  const summaryX = leftMargin + pageWidth - 200;
  yPos += 10;
  pdf.moveTo(summaryX - 20, yPos).lineTo(summaryX + 150, yPos).stroke();

  yPos += 10;
  pdf.font('Helvetica', 10);
  pdf.text('Subtotal:', summaryX, yPos);
  pdf.text(`$${data.subtotal.toFixed(2)}`, summaryX + 120, yPos, { align: 'right' });

  if (data.tax) {
    yPos += 20;
    pdf.text('Tax:', summaryX, yPos);
    pdf.text(`$${data.tax.toFixed(2)}`, summaryX + 120, yPos, { align: 'right' });
  }

  yPos += 20;
  pdf.font('Helvetica-Bold', 12);
  pdf.text('Total:', summaryX, yPos);
  pdf.text(`$${data.total.toFixed(2)}`, summaryX + 120, yPos, { align: 'right' });

  if (data.notes || data.paymentTerms) {
    yPos += 50;
    pdf.font('Helvetica-Bold', 11).text('Notes & Terms:', leftMargin, yPos);
    pdf.font('Helvetica', 10).text(data.notes || '');
    if (data.paymentTerms) pdf.text(`Payment Terms: ${data.paymentTerms}`);
  }
}
